packages/eslint-plugin/tests/rules/restrict-template-expressions.test.ts

```diff
@@ -227,6 +227,44 @@ ruleTester.run('restrict-template-expressions', rule, {
         }
       `,
     },
+    // allowNever
+    {
+      options: [{ allowNever: true }],
+      code: `
+        const arg = 'hello';
+        const msg = typeof arg === 'string' ? arg : \`arg = \${arg}\`;
+      `,
+    },
+    {
+      options: [{ allowNever: true }],
+      code: `
+        function test(arg: 'one' | 'two') {
+          switch (arg) {
+            case 'one':
+              return 1;
+            case 'two':
+              return 2;
+            default:
+              throw new Error(\`Unrecognised arg: \${arg}\`);
+          }
+        }
+      `,
+    },
+    {
+      options: [{ allowNever: true }],
+      code: `
+        // more variants may be added to Foo in the future
+        type Foo = { type: 'a'; value: number };
+        
+        function checkFoosAreMatching(foo1: Foo, foo2: Foo) {
+          if (foo1.type !== foo2.type) {
+            // since Foo currently only has one variant, this code is never run, and \`foo1.type\` has type \`never\`.
+            throw new Error(\`expected \${foo1.type}, found \${foo2.type}\`);
+            // ^ Invalid type "never" of template literal expression.
+          }
+        }
+      `,
+    },
```

Comment:
These test cases for the default case are nice and thorough. Though they all roughly boil down to concatenating something of `never` type. Which means they can be replaced with the simpler:

```ts
declare const value: never;

const stringy = `${value}`;
```

If you strongly feel a desire to have more thorough creations of `never` I won't block on it. 😄

Reply:
That makes sense! I've added a more simple test as suggested.

I think as extra security, and to provide clarity on what the allowNever option is trying to achieve I'll leave the existing tests.